src/middlewared/middlewared/plugins/nvmet/kernel.py

```diff
@@ -0,0 +1,790 @@
+import abc
+import os
+import pathlib
+import subprocess
+import time
+from collections import defaultdict
+from contextlib import contextmanager
+
+from middlewared.plugins.zfs_.utils import zvol_name_to_path
+from .constants import (DHCHAP_DHGROUP, DHCHAP_HASH, NAMESPACE_DEVICE_TYPE, NVMET_KERNEL_CONFIG_DIR,
+                        NVMET_NODE_A_ANA_GRPID, NVMET_NODE_B_ANA_GRPID, PORT_ADDR_FAMILY, PORT_TRTYPE)
+
+ANA_OPTIMIZED_STATE = 'optimized'
+ANA_INACCESSIBLE_STATE = 'inaccessible'
+ANA_PORT_INDEX_OFFSET = 5000
+
+
+class NvmetConfig(abc.ABC):
```

Comment:
I see no reason that this should be an abstract base class with a bunch of @classmethod's. This becomes incredibly inefficient because it means each method that is called creates a new instance of the class. There is no reason for that.

Marking the `post_create`, `post_update`, and `config_dict` methods as `@abstractmethod` so that classes that subclass `NvmetConfig` have to override these methods in the subclass (you're already doing this). This is the common paradigm for designing classes like this.

Otherwise, there is an incredible amount of python setup for each method being called.

Reply:
Hmmm.  I was trying to **avoid** class instantiation.

OK, I've removed the ABC and `@classmethod`.  Didn't add `@abstractmethod` because its optional as to whether they are overridden in subclasses.

The result is that each of the classes gets instantiated once per `write_config` call.  (Perhaps we could also cache these?)